File: Debugging/Mpi/Itac/Verfification/buffer_size.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main() {
    const int root = 0;
    int buff_size = 5;
    int my_rank, my_size;
    double *sbuff, *rbuff;
    MPI_Init(NULL, NULL);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &my_size);
    if (my_rank == root + 1)
        buff_size += 3;
    sbuff = (double *) malloc(buff_size*sizeof(double));
    if (my_rank == root)
        rbuff = (double *) malloc(buff_size*sizeof(double));
    for (int i = 0; i < buff_size; i++)
        sbuff[i] = pow(10.0, my_rank)*(1.0 + i);
    MPI_Reduce(sbuff, rbuff, buff_size, MPI_DOUBLE, MPI_SUM, root,
               MPI_COMM_WORLD);
    if (my_rank == root)
        for (int i = 0; i < buff_size; i++)
            printf("%d: %.2f\n", i, rbuff[i]);
    MPI_Finalize();
    return 0;
}

File: Debugging/Mpi/Itac/Verfification/buffer_size.f90
program buffer_size
      use :: mpi_f08
      implicit none
      integer, parameter :: corr_size = 10, root = 0
      integer, dimension(:), allocatable :: sbuff
      integer, dimension(corr_size) :: rbuff
      integer :: my_rank, my_size, buff_size, i

      call MPI_Init()

      call MPI_Comm_rank(MPI_COMM_WORLD, my_rank)
      call MPI_Comm_size(MPI_COMM_WORLD, my_size)

      if (my_rank == root + 1) then
          buff_size = corr_size
      else
          buff_size = corr_size
      end if
      print '("rank ", I0, ": size ", I0)', &
          my_rank, buff_size
      allocate(sbuff(buff_size))
      do i = 1, buff_size
          sbuff(i) = 10**my_rank + i
      end do

      if (my_rank /= root + 1) then
          call MPI_Reduce(sbuff, rbuff, buff_size, &
                          MPI_INTEGER, MPI_SUM, root, &
                          MPI_COMM_WORLD)
      else
          call MPI_Reduce(sbuff, rbuff, buff_size, &
                          MPI_CHARACTER, MPI_SUM, root, &
                          MPI_COMM_WORLD)
      end if

      if (my_rank == root) &
          print '(I0)', rbuff

      call MPI_Finalize()

end program buffer_size
